rcl/include/rcl/introspection.h

```diff
@@ -0,0 +1,162 @@
+// Copyright 2022 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCL__SERVICE_INTROSPECTION_H_
+#define RCL__SERVICE_INTROSPECTION_H_
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#include "rcl/publisher.h"
+#include "rcl/service.h"
+#include "rcl/time.h"
+#include "rmw/rmw.h"
+#include "rcl/client.h"
+
+#define RCL_SERVICE_INTROSPECTION_TOPIC_POSTFIX "/_service_event"
+
+typedef struct rcl_service_introspection_utils_s {
```

Comment:
if this is only usef from rcl itself and not from rclcpp/rclpy/etc, the we should move the header file to the src file (i.e. not make it public).
In that case, ignore my other comment about using pimpl.

Reply:
Parts of this header file is used from the client libraries. See: https://github.com/ros2/rclcpp/tree/service_introspection-serialized (WIP)

Reply:
If you move all the functions that act on a `rcl_service_t` directly to `service.h` that's probably not more the case already.
i.e. the functions that take `rcl_service_introspection_utils_t` as an argument are probably not being used from other client libraries (if not currently the case, I think it could be avoided).
Though I haven't checked your `rclcpp` WIP yet.